### ISSUE TITLE
error when accessing flexible array member

### ISSUE THREAD

**Issue by andihit** [iovisor/bcc#858]
Hi, when I try to access a flexible array member of a struct the following code is generated:
```
int x = ({ typeof(struct ceph_osd_req_op []) _val; memset(&_val, 0, sizeof(_val)); bpf_probe_read(...
```
which fails to compile because `typeof(struct ceph_osd_req_op []) _val;` is not valid C code.

If I'm not mistaken the following code should be produced:
```
int x = ({ typeof(struct ceph_osd_req_op *) _val; memset(&_val, 0, sizeof(_val)); bpf_probe_read(...
```
by this function: https://github.com/iovisor/bcc/blob/21dd7b46fd21eed5e0aae709a3677bc31e9817e1/src/cc/frontends/clang/b_frontend_action.cc#L248
Is this correct? If so, I could prepare a pull request.

**Comment by 4ast** [comment]
i don't think that's equivalent. cc @drzaeus77

**Comment by andihit** [comment]
you are right, they are not equivalent (I just tested it by manually patching the preprocessed C file)
any ideas?

**Comment by andihit** [comment]
The preprocessor can't read the entire array because its size is unknown to the preprocessor.
Therefore the only fix is to read ahead, check if there is an array operator after the current member expression and read only the requested array index?
In the meantime I've prepared a macro to get the address of a specific index of a flexible array member:
```
#define FLEX_ARRAY_ADDR(S, ST, F, FT, I) ((char*)S + offsetof(ST, F) + I * sizeof(FT))
struct ceph_osd_req_op op;
bpf_probe_read(&op, sizeof(op), FLEX_ARRAY_ADDR(req, struct ceph_osd_request, r_ops, struct ceph_osd_req_op, 0));
```
